R/graph_layout.R

```diff
@@ -414,3 +415,153 @@ ComputeLayout.Seurat <- function (
                "nodes in the graph"))
   }
 }
+
+
+#' Layout Coordinates Utility Functions
+#'
+#' Utility function used to manipulate layout coordinates.
+#' These functions always returns a \code{tbl_df} object.
+#'
+#' @section Center Layout Coordinates:
+#' Centers each axis of the layout coordinates around their means.
+#'
+#' @param layout A matrix-like object with layout coordinates
+#'
+#' @name layout coordinates utils
+#' @rdname layout_coordinates_utils
+#'
+#' @return A \code{tbl_df} object with adjusted layout coordinates
+#'
+#' @examples
+#' library(dplyr)
+#' library(tibble)
+#'
+#' # Generate random points that are offset to (100, 100, 100)
+#' xyz <- matrix(rnorm(600, mean = 100, sd = 20), ncol = 3,
+#'               dimnames = list(NULL, c("x", "y", "z"))) %>%
+#'   as_tibble()
+#'
+#' # Visualize random points
+#' plotly::plot_ly(data = xyz, x = ~x, y = ~y, z = ~z,
+#'                 type = "scatter3d", mode = "markers")
+#'
+#' # Center points at (0, 0, 0)
+#' xyz_centered <- center_layout_coordinates(xyz)
+#' apply(xyz_centered, 2, mean)
+#' plotly::plot_ly(data = xyz_centered, x = ~x, y = ~y,
+#'                 z = ~z, type = "scatter3d", mode = "markers")
+#'
+#' # Normalize points to have a median radius of 1
+#' xyz_normalized <- normalize_layout_coordinates(xyz_centered)
+#' radii <- sqrt(rowSums(xyz_normalized^2))
+#' median(radii)
+#' plotly::plot_ly(data = xyz_normalized, x = ~x, y = ~y,
+#'                 z = ~z, type = "scatter3d", mode = "markers")
+#'
+#' # Project points on unit sphere
+#' xyz_projected <- project_layout_coordinates_on_unit_sphere(xyz_normalized)
+#' radii <- sqrt(rowSums(xyz_projected^2))
+#' all(near(radii, y = rep(1, length(radii)), tol = 1e-12))
+#' plotly::plot_ly(data = xyz_projected, x = ~x, y = ~y,
+#'                 z = ~z, type = "scatter3d", mode = "markers")
+#'
+#' @export
+#'
+center_layout_coordinates <- function (
+  layout
+) {
+
+  stopifnot(
+    "'layout' must be a non-empty, matrix-like object" =
+      inherits(layout, what = c("matrix", "data.frame")) &&
+      length(layout) > 0,
+    "'layout' can only have 2 or 3 columns" =
+      ncol(layout) %in% c(2, 3)
+  )
+
+  # Force rename columns to x, y, z
+  colnames(layout) <- c("x", "y", "z")[1:ncol(layout)]
+  if (inherits(layout, what = "matrix")) {
+    layout <- as_tibble(layout)
+  }
+
+  # Center layout coordinates
+  layout <- layout %>%
+    mutate(across(contains(c("x", "y", "z")), ~ .x - mean(.x)))
+
+  return(layout)
+}
+
+
+#' @section Normalize Layout Coordinates:
+#' Centers each axis of the layout coordinates around the mean and
+#' adjusts each point coordinate such that their median length
+#' (euclidean norm) is 1.
+#'
+#' @param layout A matrix-like object with layout coordinates
+#'
+#' @rdname layout_coordinates_utils
+#'
+#' @export
+#'
+normalize_layout_coordinates <- function (
```

Comment:
I suggest an argument `quantile` (default 0.5 for median) that let's you decide the normalization quantile, which would make it more flexible. At some cases, it might be desirable to have almost exactly radius 1 or almost radius 1, e.g. `quantile = 0.95` will let you have a radius of almost 1 with outliers outside. 
So the change would be from  
```
median_radius <- median(radii)
layout <- layout %>%
    mutate(across(contains(c("x", "y", "z")), ~ .x/median_radius))
```
To:
```
norm_radius <- quantile(radii, quantile = quantile)
layout <- layout %>%
    mutate(across(contains(c("x", "y", "z")), ~ .x/norm_radius))
```